pkg/runner/operations/delete/delete.go

```diff
@@ -29,50 +29,64 @@ func New(client client.Client, obj ctrlclient.Object, namespacer namespacer.Name
 	}
 }
 
-func (o *operation) Exec(ctx context.Context) (_err error) {
+func (o *operation) Exec(ctx context.Context) error {
 	logger := logging.FromContext(ctx).WithResource(o.obj)
-	defer func() {
-		if _err == nil {
-			logger.Log(logging.Delete, logging.DoneStatus, color.BoldGreen)
-		} else {
-			logger.Log(logging.Delete, logging.ErrorStatus, color.BoldRed, logging.ErrSection(_err))
-		}
-	}()
+
 	if o.namespacer != nil {
 		if err := o.namespacer.Apply(o.obj); err != nil {
 			return err
 		}
 	}
+
 	logger.Log(logging.Delete, logging.RunStatus, color.BoldFgCyan)
-	candidates, _err := internal.Read(ctx, o.obj, o.client)
-	if _err != nil {
-		if errors.IsNotFound(_err) {
+
+	return o.deleteResource(ctx, logger)
+}
+
+func (o *operation) deleteResource(ctx context.Context, logger logging.Logger) error {
+	candidates, err := internal.Read(ctx, o.obj, o.client)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			logger.Log(logging.Delete, logging.DoneStatus, color.BoldGreen)
 			return nil
 		}
-		return _err
+		logger.Log(logging.Delete, logging.ErrorStatus, color.BoldRed, logging.ErrSection(err))
+		return err
 	}
+
 	for i := range candidates {
-		err := o.client.Delete(ctx, &candidates[i])
-		if err != nil && !errors.IsNotFound(err) {
+		candidate := candidates[i]
+		if err := o.tryDeleteCandidate(ctx, &candidate); err != nil {
+			logger.Log(logging.Delete, logging.ErrorStatus, color.BoldRed, logging.ErrSection(err))
 			return err
 		}
-	}
-	gvk := o.obj.GetObjectKind().GroupVersionKind()
-	for i := range candidates {
-		if err := wait.PollUntilContextCancel(ctx, internal.PollInterval, true, func(ctx context.Context) (bool, error) {
-			var actual unstructured.Unstructured
-			actual.SetGroupVersionKind(gvk)
-			err := o.client.Get(ctx, client.ObjectKey(&candidates[i]), &actual)
-			if err != nil {
-				if errors.IsNotFound(err) {
-					return true, nil
-				}
-				return false, err
-			}
-			return false, nil
-		}); err != nil {
+		if err := o.waitForDeletion(ctx, &candidate); err != nil {
+			logger.Log(logging.Delete, logging.ErrorStatus, color.BoldRed, logging.ErrSection(err))
 			return err
 		}
 	}
+
+	logger.Log(logging.Delete, logging.DoneStatus, color.BoldGreen)
```

Comment:
things like this are not correct, it should be in a defer to guarantee the log is printed